synapse/handlers/message.py

```diff
@@ -225,6 +239,89 @@ def get_joined_members(self, requester, room_id):
             for user_id, profile in iteritems(users_with_profile)
         }
 
+    @defer.inlineCallbacks
+    def maybe_schedule_next_expiry(self, expiry_ts):
+        """Schedule the expiry of an event if there's not already one running, or if the
+        one running is for an event that will expire after the provided timestamp.
+
+        Args:
+            expiry_ts (int): Timestamp to compare the time at which the next expiry will
+                happen with.
+        """
+        if not self._scheduled_expiry or not self._scheduled_expiry.active():
+            yield self._schedule_next_expiry()
```

Comment:
this is racy: you could end up with two calls to `maybe_schedule_next_expiry` happening at once, both will see `_scheduled_expiry` is unset, and end up with two tasks scheduled. (it can also race with the initial background process).

Reply:
also: it feels wasteful to go and search the table when we already know when the expiry should happen?